Python/Product/PythonTools/PythonTools/Project/PythonProjectNode.cs

```diff
@@ -777,76 +780,104 @@ internal partial class PythonProjectNode :
+        // For example, conda environments are discovered asynchronously and are
+        // not available when the project it loaded. So once they are enumerated,
+        // OnInterpreterFactoriesChanged is triggered, which calls this method.
         private void RefreshInterpreters(bool alwaysCollapse = false) {
-            if (IsClosed) {
-                return;
-            }
 
-            var node = _interpretersContainer;
-            if (node == null) {
+            // This method is re-entrant the first time it's called because GetInterpreterConfigurations() calls EnsureInitialized(),
+            // which triggers the OnInterpreterFactoriesChanged event. So only allow this method to run if it's not already running
+            if (_isRefreshingInterpreters) {
                 return;
             }
+            _isRefreshingInterpreters = true;
 
-            var remaining = node.AllChildren.OfType<InterpretersNode>().ToList();
-
-            if (!IsActiveInterpreterGlobalDefault) {
-                foreach (var fact in InterpreterFactories) {
-                    if (!RemoveFirst(remaining, n => !n._isGlobalDefault && n._factory == fact)) {
-                        bool isProjectSpecific = _vsProjectContext.IsProjectSpecific(fact.Configuration);
-                        bool canRemove = !this.IsAppxPackageableProject(); // Do not allow change python enivronment for UWP
-                        node.AddChild(new InterpretersNode(
-                            this,
-                            fact,
-                            isInterpreterReference: !isProjectSpecific,
-                            canDelete:
-                                isProjectSpecific &&
-                                Directory.Exists(fact.Configuration.GetPrefixPath()),
-                            isGlobalDefault: false,
-                            canRemove: canRemove
-                        ));
-                    }
+            try {
+
+                // if the project is closed, we're done
+                if (IsClosed) {
+                    return;
                 }
-            } else {
-                var fact = ActiveInterpreter;
-                if (fact.IsRunnable() && !RemoveFirst(remaining, n => n._isGlobalDefault && n._factory == fact)) {
-                    node.AddChild(new InterpretersNode(this, fact, true, false, true));
+
+                // if the "Python Environments" node doesn't exist, we're done
+                var pythonEnvironmentsNode = _interpretersContainer;
+                if (pythonEnvironmentsNode == null) {
+                    return;
                 }
-            }
 
-            foreach (var id in InvalidInterpreterIds) {
-                if (!RemoveFirst(remaining, n => n._absentId == id)) {
-                    node.AddChild(InterpretersNode.CreateAbsentInterpreterNode(this, id));
+                // clear out all interpreter nodes since we're going to re-add them
```

Comment:
why are all interpreter nodes removed and added again? didn't follow this part.

Reply:
Because interpreters have changed. The logic before was too convoluted. It's much easier to just clear the list in the solution explorer, and repopulated it with all the interpreters that were found, plus dummy nodes for interpreters that are referenced but no longer found, plus a global interpreter if appropriate.

This logic is much easier to follow, there's no checking what's already there, no collisions. Simple. And the performance difference is negligible because we're just updating nodes in solution explorer, and the number of interpreters is never in the hundreds or anything like that.